res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js

```diff
@@ -670,19 +670,92 @@ TraktorS2MK3.wheelDeltas = function(deckNumber, value) {
+    let targetGroups = [];
     const fxNumber = parseInt(field.id[field.id.length - 1]);
-    const group = "[EffectRack1_EffectUnit" + fxNumber + "]";
+    let noShift = false;
 
-    // Toggle effect unit
-    TraktorS2MK3.fxButtonState[fxNumber] = !TraktorS2MK3.fxButtonState[fxNumber];
+    // Target decks whose shift button is pressed.
+    for (const group of availableGroups) {
+        if (TraktorS2MK3.shiftPressed[group]) { targetGroups.push(group); }
+    }
+
+    // If no shift button was pressed fall back to target all decks.
+    if (targetGroups.length === 0) {
+        targetGroups = availableGroups;
+        noShift = true;
+    }
+
+    const fxToApply = {};
+    const activeFx = {};
+    // Detect which fx should be enabled
+    for (const group of targetGroups) {
+        activeFx[group] = engine.getValue(`[QuickEffectRack1_${group}]`, "loaded_chain_preset");
+        if (activeFx[group] === fxNumber) {
+            // Pressing again the fx button
+            fxToApply[group] = fxNumber + fxButtonCount;
+        } else if (activeFx[group] === fxNumber + fxButtonCount) {
+            // Already on secondary fx so reseting back to fxNumber
+            fxToApply[group] = fxNumber;
+        } else {
+            // First press of a different fx button
+            fxToApply[group] = fxNumber;
+        }
+    }
+
+    /* When we target both decks - No shift pressed - we want *
+     * to reset fx to same value for both */
+    if (targetGroups.length === 2 && noShift) {
+        const [deck1, deck2] = targetGroups;
+        const activeFxDeck1 = activeFx[deck1];
+        const activeFxDeck2 = activeFx[deck2];
+        const fxToApplyDeck1 = fxToApply[deck1];
+        const fxToApplyDeck2 = fxToApply[deck2];
+
+        // If any of deck has already fx enabled but different value to apply
+        if ((activeFxDeck1 === fxNumber || activeFxDeck2 === fxNumber) && fxToApplyDeck1 !== fxToApplyDeck2) {
+        // find lower value to reset both to the same one
+            const fxToApplyAll = Math.min(fxToApplyDeck1, fxToApplyDeck2);
+            fxToApply[deck1] = fxToApplyAll;
+            fxToApply[deck2] = fxToApplyAll;
+        }
+    }
 
-    engine.setValue(group, "group_[Channel1]_enable", TraktorS2MK3.fxButtonState[fxNumber]);
-    engine.setValue(group, "group_[Channel2]_enable", TraktorS2MK3.fxButtonState[fxNumber]);
-    TraktorS2MK3.outputHandler(TraktorS2MK3.fxButtonState[fxNumber], field.group, "fxButton" + fxNumber);
+    // Now apply the new fx value
+    for (const group of targetGroups) {
+        engine.setValue(`[QuickEffectRack1_${group}]`, "loaded_chain_preset", fxToApply[group]);
+    }
+};
+
+TraktorS2MK3.fxOutputHandler = function() {
+    const fxButtonCount = 4;
+    const availableGroups = ["[Channel1]", "[Channel2]"];
+
+    const activeFx = {};
+    for (const group of availableGroups) {
+        activeFx[group] = engine.getValue(`[QuickEffectRack1_${group}]`, "loaded_chain_preset");
```

Comment:
```suggestion
        activeFx[group] = engine.getValue(`[QuickEffectRack1_${group}]`, "loaded_chain_preset") - 1;
```

Reply:
Actually I didn't want to sort it out this way, because I believe there's probably a bug in the quick effects settings page.
in 2.4-beta - at least for me: 

- the list order is never saved upon Mixxx exit and needs to be changed after restart
- List index does not start with the first effect in the list but with "---" - If you get to the effect settings, hit "Restore Defaults", then the first index becomes the first effect

I believe that working around this behavior in the mapping will thus break next time the effect list behavior is sorted out

Reply:
> the list order is never saved upon Mixxx exit and needs to be changed after restart

Can't confirm. The list is saved correctly to effects.xml when Mixxx exits. Are you sure Mixxx is shutting correctly?

Reply:
> List index does not start with the first effect in the list but with "---" - If you get to the effect settings, hit "Restore Defaults", then the first index becomes the first effect
I would imagine this is a bug in the restore logic ("---" is not re-added when restoring)?

Reply:
Mmmmh I must have badly tested, you're right list is properly saved, but "---" disappear on "Restore Defaults" which would break the "workaround" mapping though.
Which way should we fix it first? :)

Reply:
I'll take a look.

Reply:
This was fixed by #11705